### PR TITLE
Adding the remoteID migration, and making the synthetic user identification more robust

### PR DESCRIPTION
**Pull request by jespino**
This should migrate the remoteID from the old users into the plugin remoteID, also ensure that there is always a remote ID, by default is msteams, but as soon as you register the remote on the shared channels it gets updated to a valid one. The reason to have a default remoteID is because not always we call the SharedChannels remote register. Probably we should do it always, and we will do eventually.

The migration is executed in every plugin restart, but in theory it is idenpotent, so it shouldn't be a problem.

### CODE REVIEW

File: server/plugin.go
Comment (jespino) [inline server/plugin.go:486]: Can you confirm if this check is needed? I would like to always register the SharedChannels for the plugin.

Comment (jespino) [reply server/plugin.go:486]: @wiggin77 :point_up:

Comment (wiggin77) [reply server/plugin.go:486]: I think we want an easy way to to turn this off in case of a problem in production.  However we don't want to use a fake remoteID either since that may be hard to clean up later.   I would suggest calling `RegisterPluginForSharedChannels` always, but then calling `UnregisterPluginForSharedChannels` if `p.getConfiguration().DisableSyncMsg == true`.  That way you will have the real remoteID.

**Review by wiggin77 — CHANGES_REQUESTED** [review]
Looks great.  I think we need the real remoteID which we can get by always registering then selectively unregistering based on config.

File: server/plugin.go
Comment (jespino) [reply server/plugin.go]: My concern is the remoteID changing over disabling/enabling. Is the remote id going to be stable or it is going to change every time that I unregister the plugin?

Comment (jespino) [reply server/plugin.go]: Anyway, I applied the changes that you proposed.

Comment (wiggin77) [reply server/plugin.go]: @jespino  You're right.  I should soft-delete the record on Unregister and clear the DeleteAt on register... or derive the remoteID from the pluginID.  Either way we want the same remoteID for every unique plugin ID.  Any preference?  I think I lend toward hashing the pluginID to make the remoteID.

Comment (jespino) [reply server/plugin.go]: Sounds good enough, but also slightly more error prone. I don't think is a big deal.

Comment (lieut-data) [reply server/plugin.go]: > Either way we want the same remoteID for every unique plugin ID.

This is probably fine for our current purposes, but thinking (very) long term, I could also see us supporting multiple Microsoft tenants concurrently which might necessitate a distinct `remoteID` for each (if we wanted to, for example, distinguish each in the UI?)

Stepping back, I'm also just wanting to confirm that /we/ (the plugin) don't really care about the value here at all, and that it's just an artifact of the way the shared channels subsystem works with users that happen to have a `RemoteID` set. In other words, we want all DMs and all GMs for all users, regardless of whether anyone has a `RemoteID` set. Does that remain accurate @wiggin77?

Comment (wiggin77) [reply server/plugin.go]: If you set the appropriate flag when registering then yes, the MS Teams plugin will get sync for all DM/GMs regardless of remoteID value.  Other plugins using the API will likely not set that flag and the remoteID on posts/users/etc will matter.  

Should we modify the register API to accept a remoteID, thus allowing a plugin to call register for each tenant?  That would also leave it up to the plugin to ensure idempotency.   cc @jespino @lieut-data

Comment (lieut-data) [reply server/plugin.go]: > Should we modify the register API to accept a remoteID, thus allowing a plugin to call register for each tenant? That would also leave it up to the plugin to ensure idempotency. cc @jespino @lieut-data

On the one hand, I don't want to preemptively build for the future. But I do like the idea of giving the plugin control over the "primary key" backing the cursor based model. At the moment, it feels just a bit too decoupled. So I'm in favour of the idea. Is the `RemoteId` constrained by some length we should remain aware of? (i.e. is it long enough for a UUID?)

Comment (wiggin77) [reply server/plugin.go]: It's the same length as most IDs in MM.

File: server/ce2e/plugin_test.go
Comment (lieut-data) [inline server/ce2e/plugin_test.go:385]: Let's constrain `(SELECT remoteId FROM remoteclusters)` to only returning the `remoteId` for this plugin, otherwise this could return multiple results and fail.

File: server/plugin.go
Comment (lieut-data) [inline server/plugin.go:498]: Propose a log here to clarify that we did indeed disable it:
```
p.API.LogInfo("Unregistering plugin for shared channels since sync msg disabled")
```

File: server/store/sqlstore/migrations.go
Comment (lieut-data) [inline server/store/sqlstore/migrations.go:10]: To confirm, all users we've ever created have had a non-nil, non-empty `RemoteID`?

File: server/testutils/containere2e/containere2e.go
Comment (lieut-data) [inline server/testutils/containere2e/containere2e.go:120]: @wiggin77, can you remind me where things stand in terms of rolling this out to production? Will we require customers to enable this experimental setting (and the enduser functionality) before we can rely on shared channels, or is there a ticket to track decoupling these?

**Review by lieut-data — APPROVED** [review]
Thanks, @jespino! There's one blocker below, but it's in the unit tests, otherwise just a handful of non-blocking questions so I'm approving to unblock.

File: server/ce2e/plugin_test.go
Comment (jespino) [reply server/ce2e/plugin_test.go:385]: As it is part of the plugin test, my expectation is there is nothing else there, I can add it, but it shouldn't be a problem.

File: server/plugin.go
Comment (jespino) [reply server/plugin.go:498]: Good call

File: server/store/sqlstore/migrations.go
Comment (jespino) [reply server/store/sqlstore/migrations.go:10]: All the synthetic users, by definition, are the ones that have `RemoteID` different that nil/"", so. Not "all the users" that we have created, the bot user, and any promoted user has no `RemoteID`, but that is the expected behavior.

File: server/testutils/containere2e/containere2e.go
Comment (jespino) [reply server/testutils/containere2e/containere2e.go:120]: Let me remove that, I think I added there while I was trying to figure out the how to fix the test while I was working on the PR, but if I can remove that line and keep going, works for me.

File: server/store/sqlstore/migrations.go
Comment (lieut-data) [reply server/store/sqlstore/migrations.go:10]: Awesome, thanks for confirming! (Mostly was just thinking back to early deployments with any customers and making sure there wasn't a time when `RemoteID` was left `nil`.)

File: server/plugin.go
Comment (lieut-data) [inline server/plugin.go:501]: nit: can we put this before we make the call? (If it fails and we log a warning, the logs won't quite make sense.)

**Review by lieut-data — APPROVED** [review]

File: server/plugin.go
Comment (jespino) [reply server/plugin.go:501]: :+1:

**Review by wiggin77 — APPROVED** [review]
LGTM :+1:

**Pull request merged**
